bin/runfx.js

```diff
@@ -0,0 +1,100 @@
+"use strict";
+
+/* eslint-env node */
+
+var Commander = require("commander");
+var Fs = require("fs-promise");
+var Path = require("path");
+var Rimraf = require("rimraf");
+var Util = require("./run-utils");
+var Version = require("../package.json").version;
+var When = require("when");
+var WhenNode = require("when/node");
+// Transform mkdirp() to use promises.
+var Mkdirp = WhenNode.lift(require("mkdirp"));
+
+var DEFAULT_PROFILE = "loop-dev";
+
+Commander
+  .version(Version)
+  .option("-b, --binary <path>", "Path of Firefox binary to use.")
+  .option("-p, --profile <path>", "Path or name of Firefox profile to use.")
+  .option("--binary-args <CMDARGS>", "Pass additional arguments into Firefox.")
+  .parse(process.argv);
+
+function ensureRemoved(path) {
+  return When.promise(function(resolve, reject) {
+    Fs.lstat(path).then(function(targetStat) {
+      if (targetStat.isDirectory()) {
+        Rimraf(path, function(err) {
+          if (err) {
+            reject("Removing add-on directory '" + path + "' failed " + err);
+            return;
+          }
+
+          resolve();
+        });
+      } else {
+        // Removing old symlink or file.
+        Fs.unlink(path).then(resolve).catch(reject);
+      }
+    }).catch(resolve);
+  });
+}
+
+function onExit() {
+  var i = 0;
+  var errLen = arguments.length;
+  for (; i < errLen; ++i) {
+    console.error(arguments[i].message || arguments[i]);
+  }
+
+  if (errLen) {
+    process.exit(1);
+  }
+}
+
+var addonSourceDir = Path.normalize(Path.join(__dirname, "..", "built", "add-on"));
+Fs.stat(addonSourceDir).then(function(sourceStat) {
+  if (!sourceStat.isDirectory()) {
+    throw "Not a directory";
+  }
+
+  // Add-on directory exists, so now we can create the symlink in the profile
+  // directory and run Firefox.
+  var profile = Commander.profile || DEFAULT_PROFILE;
+
+  Util.getProfilePath(profile).then(function(profilePath) {
+    // Since we've got the profile path now, we can create the symlink in the
+    // profile directory IF it doesn't exist yet.
+    var extensionsDir = Path.join(profilePath, "extensions");
+    return Mkdirp(extensionsDir).then(function() {
+      var addonTargetDir = Path.join(extensionsDir, "loop@mozilla.org");
+      return ensureRemoved(addonTargetDir).then(function() {
+        // This should fail at a certain point, because we don't want the add-on
+        // directory to exist.
+        return Fs.symlink(addonSourceDir, addonTargetDir).then(function() {
+          console.log("Symlinked add-on at '" + addonTargetDir + "'");
+
+          // Add-on should be in the correct place, so now we can run Firefox.
+          return Util.runFirefox(Commander).then(function() {
+            throw "Firefox terminated properly";
+          }).catch(function() {
+            console.log("Removing symlink");
+            Fs.unlink(addonTargetDir);
+          });
+        });
+      }).catch(onExit);
+    }).catch(onExit);
```

Comment:
There's a lot of nesting and hard-to-match-catches that seems unnecessary given the use of Promises. I made a commit that should be pretty much equivalent: https://github.com/Mardak/loop/commit/1de36d268f3fed283638c2be11572b7a5e2044e8